### PR TITLE
feat: External links download files

### PR DESCRIPTION
**Pull request by adrien-coye**
- [x] Add to my drive
- [x] Download All
- [x] New matomo tags

root PR
[List files](https://github.com/Infomaniak/ios-kDrive/pull/1307)

Sub PR
[UpSale sheet](https://github.com/Infomaniak/ios-kDrive/pull/1309)
[Multi select](https://github.com/Infomaniak/ios-kDrive/pull/1311)

### CODE REVIEW

File: kDrive/UI/Controller/Files/File List/FileListViewController.swift [hunk 1]
```diff
@@ -250,6 +252,46 @@ class FileListViewController: UICollectionViewController, SwipeActionCollectionV
         }
     }
 
+    func setupFooterIfNeeded() {
+        guard driveFileManager.isPublicShare else {
+            return
+        }
+
+        let addToKDriveButton = IKButton(type: .custom)
+        addToKDriveButton.setTitle("Add to My kDrive", for: .normal)
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Files/File List/FileListViewController.swift:261]: i18n

File: kDrive/UI/Controller/Files/File List/FileListViewController.swift [hunk 2]
```diff
@@ -348,6 +390,30 @@ class FileListViewController: UICollectionViewController, SwipeActionCollectionV
         }
     }
 
+    private func fileLayout(files: [File]) -> FloatingPanelLayout {
```
Comment (PhilippeWeidmann) [inline kDrive/UI/Controller/Files/File List/FileListViewController.swift:393]:
```suggestion
    private func fileFloatingPanelLayout(files: [File]) -> FloatingPanelLayout {
```

File: kDrive/UI/Controller/Files/File List/FileListViewController.swift [hunk 3]
```diff
@@ -896,3 +958,38 @@ extension FileListViewController: UICollectionViewDropDelegate {
...
+    func setBackgroundColors(normal normalColor: UIColor, highlighted highlightedColor: UIColor) {
+        if let normalImage = UIImage(color: normalColor) {
+            setBackgroundImage(normalImage, for: .normal)
+        }
+
+        if let highlightedImage = UIImage(color: highlightedColor) {
+            setBackgroundImage(highlightedImage, for: .highlighted)
+        }
+    }
+}
```
Comment (PhilippeWeidmann) [inline kDrive/UI/Controller/Files/File List/FileListViewController.swift:995]: Is it still needed with IKLargeButton ?

**Review by PhilippeWeidmann — COMMENTED** [review]
Nice work ! Left some comments mostly regarding the footer button

File: kDrive/UI/Controller/Files/File List/FileListViewController.swift [hunk 4]
```diff
@@ -250,6 +252,46 @@ class FileListViewController: UICollectionViewController, SwipeActionCollectionV
         }
     }
 
+    func setupFooterIfNeeded() {
+        guard driveFileManager.isPublicShare else {
+            return
+        }
+
+        let addToKDriveButton = IKButton(type: .custom)
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/UI/Controller/Files/File List/FileListViewController.swift:260]: I don't see the value of the lazy pattern here. I did put the button factory behind a var for the sake of splitting the code.

File: kDrive/UI/Controller/Files/File List/FileListViewController.swift [hunk 5]
```diff
@@ -348,6 +390,30 @@ class FileListViewController: UICollectionViewController, SwipeActionCollectionV
...
+    private func fileFloatingPanelLayout(files: [File]) -> FloatingPanelLayout {
+        guard driveFileManager.isPublicShare else {
+            return FileFloatingPanelLayout(
+                initialState: .half,
+                hideTip: true,
+                backdropAlpha: 0.2
+            )
+        }
+
+        if files.first?.isDirectory ?? false {
```
Comment (PhilippeWeidmann) [inline kDrive/UI/Controller/Files/File List/FileListViewController.swift:402]:
```suggestion
        if files.first?.isDirectory == true {
```

File: kDriveCore/Data/Api/DriveApiFetcher.swift [hunk 6]
```diff
@@ -59,6 +59,10 @@ public struct PublicShareMetadata: Decodable {
 
     public let validUntil: TimeInterval?
     public let capabilities: Rights
+    
+    // TODO: Test parsing
+    public let isPasswordNeeded: Bool = false
+    public let isExpired: Bool = false
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/Api/DriveApiFetcher.swift:65]: Also not sure this will work. At which point do you want to test it ?

Comment (adrien-coye) [reply kDriveCore/Data/Api/DriveApiFetcher.swift:65]: Password protected will be in another PR, as well as expired.
I made sure to add completion steps to the feature PR.

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**